specification/dartLangSpec.tex

```diff
@@ -16231,238 +16306,447 @@ \subsection{Imports}
+If a call fails, the library has not been loaded,
+and one has the option to invoke \code{$p$.loadLibrary()} again.
+Whether a repeated call to \code{$p$.loadLibrary()} succeeds will vary,
+as described below.
 
-In contrast a library comprises both imports and their usage; the library is under the control of a single party and so any problem stemming from the import can be resolved even if it is reported at the use site.%
+\commentary{%
+Note that it is a compile-time error for a deferred prefix
+to be used in more than one import,
+which means that the update of the binding of $p$
+does not interfere with other imports.%
 }
 
-\commentary{
-Note that no errors are raised if one hides or shows a name that is not in a namespace.
-}
-\rationale{
-This prevents situations where removing a name from a library would cause breakage of a client library.
+\LMHash{}%
+The effect of a repeated invocation of \code{$p$.loadLibrary()} is as follows:
+\begin{itemize}
+\item
+  If another invocation of \code{$p$.loadLibrary()} has already succeeded,
+  the repeated invocation also succeeds.
+  Otherwise,
+\item
+  If another invocation of \code{$p$.loadLibrary()} has failed:
+  \begin{itemize}
+  \item
+    If the failure is due to a compilation error,
+    the repeated invocation fails for the same reason.
+  \item
+    If the failure is due to other causes,
+    the repeated invocation behaves as if no previous call had been made.
+  \end{itemize}
+\end{itemize}
+
+\commentary{%
+In other words, one can retry a deferred load after a network failure
+or because a file is absent,
+but once one finds some content and loads it, one can no longer reload.
+We do not specify which object the returned future resolves to.%
 }
+\EndCase
 
 \LMHash{}%
-The dart core library \code{dart:core} is implicitly imported into every dart library other than itself via an import clause of the form
+\Case{Semantics of immediate imports}
+When $I_i$ is an immediate import that refers to a URI via the string $s_i$:
+Let $\ell_i$ denote the run-time entity that corresponds to the result of
+compiling the library $L_i$ denoted by $s_i$.
 
-\code{\IMPORT{} `dart:core';}
+\LMHash{}%
+$\ell_i$ is accessible in the current isolate at run time
+due to an implementation specific mechanism,
+and it provides access to the result of compiling functions,
+as well as other resources needed for the execution of code in $L_i$.
 
-unless the importing library explicitly imports \code{dart:core}.
+\LMHash{}%
+Let \NamespaceName{\metavar{import}, i} be the namespace imported from $L_i$.
+The run-time namespace \NamespaceName{i} will then have the following bindings:
 
-\commentary{
-Any import of \code{dart:core}, even if restricted via \SHOW{}, \HIDE{} or \AS{}, preempts the automatic import.
-}
+\begin{itemize}
+\item
+  For every top level function, getter, or setter $m$ named $n$ in
+  \NamespaceName{\metavar{import}, i},
+  a binding from $n$ to the result of compiling $m$,
```

Comment:
I don't even think we should mention *compiling* in the semantics. That too is an implementation detail.

So, just:
> a binding from *n* to the declaration named *n* exported by *&ell<sub>i<sub>*;

Dart source has a semantics. The spec is defining the semantics of such source code, at a sufficiently abstract level that a compiler can transform the source code into something that can be run with that semantics.

We do that by association domain entities with source entities, so a *class declaration* (source) defines a *class* (abstract semantic entity with certain characteristics that we also define).
We can define both as static and a dynamic semantics, but for classes and libraries, there doesn't have to be a difference, because both are in one-to-one correspondence with the source declaration. (Arguably, static and library variables muddy the picture, but we can pretend the static semantics just ignore the state of those). 
For things with dynamic behavior (anything with statements and non-constant expressions), we need to specify static and dynamic semantics separately, but for declarations, we don't need to, so we shouldn't.
(Which is again why there is no reason to distinguish a library from a "runtime entity" representing the library, there is only one library).

Reply:
True, the program could be interpreted, and we don't want to imply that it isn't.

But 'compilation' would also be a reasonable reference to the process whereby an interpreter creates a program representation that enables interpretation. In practice it will create bytecode to represent expressions/statements, trees to represent inheritance structures, etc.

I think the claim that there is a `declaration` at run time is more likely to create confusion, and will probably be taken as a hint that we are unaware of the distinction between compile time and run time.

So do you really think it's a negative element to have this minimal reference to the existence of a process (let's call it `compilation`) that makes programs runnable?

We certainly can't insist that there is a 1:1 correspondence between source code and run time; for instance, source code doesn't include implicitly induced getters and setters; or default constructors; or noSuchMethod forwarders; etc.

Reply:
So far we have always talked about declarations at run-time. We traverse class hierarchies to find member declarations when we do member invocations. Then we invoice the method, which executes the body in a scope where we bind the *actual* `this` and `super`. That's worked fine so far. We don't need a run-time only representation of the concept of a method declaration, because we assign semantics to running the body's *syntax* anyway. The variables need to have run-time bindings though.

Reply:
I removed every reference to compilation, cf. the other comment on the same topic.